pkg/ottl/functions.go

```diff
@@ -77,9 +77,16 @@ func (p *Parser[K]) buildArgs(ed editor, argsVal reflect.Value) error {
 		argVal := ed.Arguments[argNum]
 
 		var val any
-		if fieldType.Kind() == reflect.Slice {
+		switch {
+		case strings.HasPrefix(fieldType.Name(), "FunctionGetter"):
+			f, ok := p.functions[string(*argVal.Enum)]
```

Comment:
I believe we need a grammar change instead of depending on Enum.  Using Enum restricts FunctionGetters to only functions with capital letters and numbers.  We need to instead allow FunctionGetter to use any Converter.

I believe we need to add something to `value` that will match the Converter syntax when a Converter symbol is used and not followed by `(`.  I don't think we to add it to `mathExprLiteral` because I don't believe a Function parameter should be allowed in mathematical expressions.

You might be able to reuse `converter` and make the arguments optional, but it will probably be easier to introduce a new struct that uses the same `Function` field as `converter` but doesnt have args or keys.

Reply:
do we want to make the grammer change in this PR/issue? It seems reasonable to start with capital letters and numbers and expand the grammar in another PR?

Reply:
Yes we need to make the grammar change in this PR in order to allow this new capability.  Adding new core features to OTTL like this one need to be well rounded so allowing only functions that match the Enum pattern is too restricting.  We don't want users or new function creators to have to worry about that kind of limitation.

Reply:
it looks like there could be a conflict between converter names and Enums if we use the above approach unless I precede the converter name with a special character (~ , $ etc..)?

Reply:
However preceding the converter name with a special char seems to be invalid syntax according to the current lexer rules. I tried adding this below to the lexer rule but that didn't seem to work.
```
{Name: `Convertercase`, Pattern: `(~)[A-Z][A-Za-z0-9_]*`}
```

Reply:
I think you're right that there will be a conflict.  I was hoping to do:

```
type value struct {
	IsNil          *isNil           `parser:"( @'nil'"`
	Literal        *mathExprLiteral `parser:"| @@ (?! OpAddSub | OpMultDiv)"`
	MathExpression *mathExpression  `parser:"| @@"`
	Bytes          *byteSlice       `parser:"| @Bytes"`
	String         *string          `parser:"| @String"`
	Bool           *boolean         `parser:"| @Boolean"`
	Enum           *EnumSymbol      `parser:"| @Uppercase"`
	FunctionName   *string          `parser:"| @(Uppercase(Uppercase | Lowercase)*)"`
	List           *list            `parser:"| @@)"`
}
```

but that will capture the function name into Enum if it is all Uppercase. Adding another identifying character to the start of FunctionName would solve this but what would we use?

Maybe the grammar shouldn't have allowed Coverters names to be all uppercase like Enums :/ 

@evan-bradley what do you think?

Reply:
There _might_ be something super fancy we can do with custom capture logic where we check if the incoming value is a function name and if it is we skip trying to match to a enum.  I'd have to go look at the participle library to be sure.

Reply:
> Adding another identifying character to the start of FunctionName would solve this but what would we use?
I was thinking maybe `SHA256()` or `$SHA256`, but the current lexer rules don't seem to support it and adding below didn't seem to work. Maybe there's something I don't understand about Participle, looking into it.
```{Name: `Convertercase`, Pattern: `[A-Z][A-Za-z0-9_]*"()"`}```

Reply:
We should not need to mess with the parser's set of tokens.  If you want to experiment with a preceeding character then this should work:

```go
type value struct {
	IsNil          *isNil           `parser:"( @'nil'"`
	Literal        *mathExprLiteral `parser:"| @@ (?! OpAddSub | OpMultDiv)"`
	MathExpression *mathExpression  `parser:"| @@"`
	Bytes          *byteSlice       `parser:"| @Bytes"`
	String         *string          `parser:"| @String"`
	Bool           *boolean         `parser:"| @Boolean"`
	Enum           *EnumSymbol      `parser:"| @Uppercase"`
	FunctionName   *string          `parser:"| '$' @(Uppercase(Uppercase | Lowercase)*)"`
	List           *list            `parser:"| @@)"`
}
```

Reply:
I assume you mean
```
FunctionName   *string          `parser:"| '$' @(Uppercase(Uppercase | Lowercase)*)"`
```
with `$` included in `Punct`, it seems to break the parsing of existing statements
```
unable to parse OTTL statement "merge_maps(attributes, body, \"insert\")": statement has invalid syntax: 1:22: invalid input text ", body, \"insert\"..."
```

Reply:
Ya I meant to use `&` in both places

Reply:
I know it's little odd, but I think we should use the `value` struct @TylerHelmuth originally suggested, and just check whether the function name got parsed into the `Enum` or `FunctionName` fields. By the time we're reading the value from the AST, we already know we're looking for a function, so it's just a matter of which field it's in.

Reply:
I pushed changes to support passing in literal converter names with a `&` prefix.

Reply:
So with this change any uppercase string prefixed with `&` is a `FunctionName` and `Enum` otherwise

- `SHA256` - Enum
- `&SHA256` - FunctionName

@evan-bradley `FunctionName` is defined in the `value` struct

Reply:
> By the time we're reading the value from the AST, we already know we're looking for a function, so it's just a matter of which field it's in

Actually it looks like it will be in either `Enum` or `FunctionName` depending on what matches first.

Reply:
> By the time we're reading the value from the AST, we already know we're looking for a function, so it's just a matter of which field it's in

So should we not go the prefix route an just check for `FunctionName` before considering it an `Enum`?

Reply:
another approach - no changes to the grammar, just check for a `FunctionName` before considering it an `Enum`. So if the `FunctionName` is a converter, then use it like so, else (unknown function?) using it as an `Enum`.

Reply:
>I believe this PR does this by introducing a grammar change to support the `&` prefix.

Sorry, I think my comment circled around my point a little bit. I'd prefer not to prefix the function name with anything when using it as a value. The set of valid enum names is a subset of the set of valid converter names, so we can rely on the fact that converters will match to `Enum` or `FunctionName`, while enums will only match to `Enum`. If we define the grammar like so:

```golang
type value struct {
	IsNil          *isNil           `parser:"( @'nil'"`
	Literal        *mathExprLiteral `parser:"| @@ (?! OpAddSub | OpMultDiv)"`
	MathExpression *mathExpression  `parser:"| @@"`
	Bytes          *byteSlice       `parser:"| @Bytes"`
	String         *string          `parser:"| @String"`
	Bool           *boolean         `parser:"| @Boolean"`
	Enum           *EnumSymbol      `parser:"| @Uppercase"`
	FunctionName   *string          `parser:"| @(Uppercase(Uppercase | Lowercase)*)"`
	List           *list            `parser:"| @@)"`
}
```

Then function names like `SHA256` will match to `Enum`, while names like `FNV1a` will match to `FunctionName`. So the function names may match to different fields on the `value` node in the AST, but we can handle this since are already know the value is intended to be a function name:

```golang
var name string

switch {
case argVal.Enum != nil:
	name = string(*argVal.Enum)
case argVal.FunctionName != nil:
	name = *argVal.FunctionName
default:
	return errors.New("invalid function name given")
}

f, ok := p.functions[name]
```

I think this would achieve what we're looking for without muddying things up too much.

>another approach - no changes to the grammar, just check for a `FunctionName` before considering it an `Enum`. So if the `FunctionName` is a converter, then use it like so, else (unknown function?) using it as an `Enum`.

I think we will need a grammar change no matter what. The regex for `Enum` doesn't capture all function names, while the regex for converter names will catch invalid `Enum` names. We could take advantage of the `checkForCustomError` functions to do this check, but I'm a little uneasy about mixing parsing the AST with turning the AST into an executable statement. I may be misreading your comment though, so please correct me if I got something wrong.

Reply:
got it, if a field of type `FunctionGetter` isn't an `Enum` or a `FunctionName` then its invalid, otherwise proceed. So `FunctionName` removes the previous limitation on converters that could be passed.

Reply:
Yeah, exactly.

Reply:
One thing to look out for is that since a FunctionName like `SHA256` matches `Enum`, `newGetter` will attempt to pass the value of the `Enum` to `p.enumParser(val.Enum)`.  We'd just have to ignore that error until we also try it as a function name.

Reply:
That's a good callout, but currently it's not an issue since the code bypasses `newGetter`. It may make sense to move this to `newGetter` since (I think) that would enable higher-order functions and allow Converters to return `FunctionGetter`s. The caveat is that we'd need to combine that with a change to check that a given context doesn't have any naming conflicts between enums and Converters. We'd also probably have to figure out how to return a meaningful error to users since unless we know the Getter type declared by the function factory, we have no way of knowing whether the user entered an invalid function name or enum name.

Reply:
yes, the code currently bypasses`newGetter` so the previous suggestion of simply checking whether a field of type `FunctionGetter` is an `Enum` or a `FunctionName` before getting its name using the corresponding field should suffice. Should I spend some time moving the code to `newGetter`?

Reply:
Let's continue to skip `newGetter` and try Evan's idea

Reply:
Also, the return type of `newGetter` is `Getter[K]`. Moving the `FunctionGetter` logic to `newGetter` might mean having to return `Factory[K]`, as `Factory[K]` is required to initialize the `StandardFunctionGetter`. The underlying reason is that a `StandardFunctionGetter` is simply initialized differently than other `Getters`. So it does seem like this code doesn't belong to `newGetter`.

Reply:
> Let's continue to skip `newGetter` and try Evan's idea

ok I have added the code below

```
switch {
case argVal.Enum != nil:
	name = string(*argVal.Enum)
case argVal.FunctionName != nil:
	name = *argVal.FunctionName
default:
	return errors.New("invalid function name given")
}
```

but looks like it broke an OTTL statement test

Reply:
nvm fixed it, needed to ensure that `Enum` is given higher priority in the grammar.

Reply:
@TylerHelmuth @evan-bradley all tests seem to be passing with the latest suggestions. Please let me know if there's anything else. Not sure if there's any updates needed to the documentation for these changes.

Reply:
@TylerHelmuth @evan-bradley looks like all tests pass with the latest suggestions. Please let me know if there's anything else including doc changes.